[Lens] Workspace panel dimensions by chart type

drewdaemon opened the pull request:
## Summary

Close https://github.com/elastic/kibana/issues/136993

All charts remain the same except the following

## Metric
Each tile gets `200px` if there are multiple and `300px` if it's a single. The default background is EUI empty.

<img width="600" alt="Screenshot 2023-10-30 at 3 55 33 PM" src="https://github.com/elastic/kibana/assets/315764/74d7e6dc-c90a-4f60-bf56-94ab1556ad42">

<img width="600" alt="Screenshot 2023-10-30 at 3 56 36 PM" src="https://github.com/elastic/kibana/assets/315764/3254160a-b18a-4c04-b059-f20b8f1f246a">

## XY
Vertical time-series charts have `16:9` ratio but will stretch to any available width and won't shrink below 300px height.

https://github.com/elastic/kibana/assets/315764/3e056ae8-bc65-4851-9ad9-6c8a81bdf58a




## Gauge

Gauge gets `600px` for the long side, `300px` for the short side.

<img width="600" alt="Screenshot 2023-11-28 at 11 22 20 AM" src="https://github.com/elastic/kibana/assets/315764/2b774515-f060-4c26-a0aa-efdeebfff0e5">

<img width="600" alt="Screenshot 2023-11-28 at 11 22 33 AM" src="https://github.com/elastic/kibana/assets/315764/62181021-d88a-4cb6-862b-42768a2df13e">



## Known issues
- [ ] text resizing on the metric https://github.com/elastic/elastic-charts/issues/2238

https://github.com/elastic/kibana/assets/315764/0162f461-b544-44a9-971c-b2a0265d7d3a

- [x] gauge isn't showing veil on willRender

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

Flaky test runner: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4755

drewdaemon commented:
@elasticmachine merge upstream

drewdaemon commented:
@elasticmachine merge upstream

drewdaemon commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

drewdaemon commented:
/ci

drewdaemon commented:
/ci

drewdaemon commented:
/ci

drewdaemon commented:
/test

drewdaemon commented:
/ci

drewdaemon commented:
/ci

drewdaemon commented:
/ci

drewdaemon commented:
/ci

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

drewdaemon commented:
@elasticmachine merge upstream

jbudz reviewed (approved):
limits.yml

nreese commented on x-pack/plugins/maps/public/lens/choropleth_chart/expression_renderer.tsx:
```diff
@@ -92,6 +93,19 @@ export function getExpressionRenderer(coreSetup: CoreSetup<MapsPluginStartDepend
         handlers.done();
       };
 
+      const dimensionsEvent: DimensionsEvent = {
```
Can you explain this a little more? Region map does not display a percentage, but rather any metric aggregation. Not sure I understand setting dimensions for a map.

drewdaemon replied:
Dimensions here refers to the size of the Lens workspace. The expression renderers can now request ideal dimensions / aspect ratio from the Lens editor. For example, horizontal bar charts can now request a 16:9 ratio.


https://github.com/elastic/kibana/assets/315764/facb0f99-6a05-4dbf-aead-896ca84ca7da

In the case of the chloropleth, we want it to act as it does today (fill all the available space). That's the reason the renderer needs to request 100% for both width and height.

mbondyra replied:
something to consider based on this comment and my preference (but of course the decision is up to you Drew) - we use the word 'dimension' in Lens with many meanings. For example, if you search for chart dimension in google you get this answer:

> What is a chart dimension?
> Defining the dimensions of a chart - Support - Bizzdesign Support
> The dimensions of a chart are used to determine what object information will be shown in the chart. Depending on the type of chart this information: is placed along one or more axes (in the bar chart for example)


We call dimension exactly that and I think that we should reconsider if we should also use the same word for chart size. In my opinion we should stick to `chartSize` instead.

drewdaemon replied:
Sure. Had this same thought when Nathan asked his question 👍

drewdaemon replied:
Done.

nreese reviewed (approved):
kibana-gis changes LGTM
code review only

stratoula reviewed (commented):
This looks so so so awesome 😍 

- I see that a FT fails. Is it flaky? Have we run the flaky test runner to ensure that we are not going to merge a flaky test?
- This is mostly a question and not an objection. I see that the XY charts have this 16:9 ratio regardless if it is a timeseries or not. I had understood that we are following this ratio only for timeseries. Have we decided against it?

mbondyra commented on src/plugins/chart_expressions/expression_xy/public/components/__snapshots__/xy_chart.test.tsx.snap:
🙈 😅

mbondyra commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/workspace_panel/workspace_panel_wrapper.tsx:
```diff
@@ -199,10 +237,30 @@ export function WorkspacePanelWrapper({
+          alignItems="center"
+          justifyContent="center"
+          direction="column"
+          css={{ height: '100%' }}
```
nit: could we instead use the emotion syntax? It's more performant this way
```suggestion
          css={css`
            height: 100%;
          `}
```

drewdaemon replied:
Actually, I see now that it uses emotion either way. See [the docs](https://emotion.sh/docs/css-prop#object-styles)

mbondyra replied:
nice! TIL 😁

stratoula commented on src/plugins/chart_expressions/common/dimension_transition_veil.tsx:
```diff
@@ -0,0 +1,62 @@
+ *
+ * It is only relevant if client dimensions are being requested based on chart configuration.
+ */
+export function useDimensionTransitionVeil(
```
@drewdaemon how can we test this?

stratoula replied:
Dismiss! @mbondyra helped me to find the veil!

mbondyra commented on src/plugins/chart_expressions/common/types.ts:
```diff
@@ -26,3 +27,29 @@ export type MakeOverridesSerializable<T> = {
+
+export interface DimensionsEvent extends ExpressionRendererEvent {
+  name: 'dimensions';
+  data: ChartDimensionOptions;
```
as my previous comment, how about `ChartSizeSpec`?

mbondyra commented:
Can we do something about horizontal bar? It happens quite often with a bit longer labels that the chart doesn't look good :( 
<img width="978" alt="Screenshot 2024-01-03 at 14 19 29" src="https://github.com/elastic/kibana/assets/4283304/ddfe6756-c35b-47d1-90e8-2a95f031956c">

drewdaemon commented on src/plugins/visualizations/public/embeddable/visualize_embeddable.tsx:
```diff
@@ -477,6 +478,9 @@ export class VisualizeEmbeddable
       this.handler.events$
         .pipe(
           mergeMap(async (event) => {
+            if (isDimensionsEvent(event)) {
```
The renderers now dispatch sizing events as well as filter events. That code originally assumed that all events were filter events. Since the legacy visualizations use the same renderers, they were incorrectly treating sizing events as filters, locking up the filtering system. Now, they just ignore non-filter events.

stratoula replied:
Got it, can you add a comment though?

drewdaemon replied:
Done.

drewdaemon commented:
> @drewdaemon in the description of this PR, could you sum up with one sentence for each chart what to expect? I know there's a discussion in the issue but it'd be good to have a summary of what behavior changes were done when testing 🙏🏼 For example, for xy chart, I am not sure if we always want to scale to 100% of available space or keep aspect ratio as it changes when I resize.

@mbondyra thanks for taking a look—the screenshots show what to expect for each chart, but I've added a bit of extra explanation.

> For example, for xy chart, I am not sure if we always want to scale to 100% of available space or keep aspect ratio as it changes when I resize.

I'm not sure I understand this piece of feedback. Perhaps we could sync.

drewdaemon commented:
@stratoula

> I see that a FT fails. Is it flaky? Have we run the flaky test runner to ensure that we are not going to merge a flaky test?

Yeah, I ran it yesterday and it's definitely flaky.

> This is mostly a question and not an objection. I see that the XY charts have this 16:9 ratio regardless if it is a timeseries or not. I had understood that we are following this ratio only for timeseries. Have we decided against it?

Good question. I didn't consider changing the dimensions based on anything but chart type, but I'm sure we could. Not sure where to discuss this.

drewdaemon commented:
@mbondyra 

> Can we do something about horizontal bar? It happens quite often with a bit longer labels that the chart doesn't look good :( 

Thanks for reminding me of this issue. Early on, the designers suggested excluding the labels from the width calculation. That makes sense in theory, but we'll have to look into how to do it technically speaking. Maybe there are other things we can do 🧠

MichaelMarcialis commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/workspace_panel/workspace_panel_wrapper.tsx:
```diff
@@ -199,10 +237,30 @@ export function WorkspacePanelWrapper({
         className={classNames('lnsWorkspacePanelWrapper stretch-for-sharing', {
           'lnsWorkspacePanelWrapper--fullscreen': isFullscreen,
         })}
+        css={{ height: '100%' }}
```
Should we use Emotion here as well?

drewdaemon replied:
the `css` prop is provided by Emotion, so we're already using it

MichaelMarcialis commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/workspace_panel/workspace_panel_wrapper.tsx:
```diff
@@ -199,10 +237,30 @@ export function WorkspacePanelWrapper({
+          <EuiFlexItem
+            data-test-subj="lnsWorkspacePanelWrapper__innerContent"
+            grow={false}
+            css={{
+              flexGrow: 0,
+              height: '100%',
+              width: '100%',
+              ...visDimensionsCSS,
+            }}
```
Same question here regarding use of Emotion.

drewdaemon replied:
the `css` prop is provided by Emotion, so we're already using it

MichaelMarcialis reviewed (commented):
Thanks for putting this together, @drewdaemon. It's looking great! I've left a few comments below for your review. Let me know if you have any questions or need any design support.

<hr/>

> long labels look bad for horizontal charts

> Can we do something about horizontal bar? It happens quite often with a bit longer labels that the chart doesn't look good :(

Unfortunately this is not limited to horizontal bar visualizations (gauges, heat maps, tree maps, and tag clouds all have issues with excessively long titles and/or labels). And while this issue is more prevalent with the introduction of aspect ratios for different visualization types, these issues already exist to some degree today.

![CleanShot 2024-01-03 at 15 21 10](https://github.com/elastic/kibana/assets/3884767/90dfb107-4658-4a6b-9f24-ff53b66c48f3)

![CleanShot 2024-01-03 at 11 18 42](https://github.com/elastic/kibana/assets/3884767/1527deba-e262-4f51-9ef6-c0dd802b7bbe)

This is a complex issue, but two potential solutions come to mind:

1. **Support line breaks and/or truncation.** Am I correct in assuming Elastic charts doesn't currently support the ability for titles and labels to break lines or be truncated? If not, can and should we support it for these sorts of situations?

2. **Change where the aspect ratio is applied.** The aspect ratio is being applied to the entire visualization container currently, meaning both the visualization and its accompanying labels, legends, and any other external accoutrements are being included in that ratio. Should the aspect ratio only be applied to the chart portion (i.e. excluding labels, legends, etc. from being part of the aspect ratio)?

Thoughts? CCing @gvnmagni and @markov00, in case they have other ideas.

<hr/>

When shrinking your browser viewport width while using a vertical bar visualization, just prior to hitting the small breakpoint threshold, I noticed the workspace isn't shrinking horizontally and is instead overflowing horizontally out of its container. Can we fix this?

![CleanShot 2024-01-03 at 14 55 35](https://github.com/elastic/kibana/assets/3884767/9f06b0f1-ed3c-424c-bb45-fc02877d6e52)

<hr/>

It appears that a square aspect ratio is being applied to table visualizations. Is this intentional? I think we should have tables always expand to the full available width and height in the workspace. Thoughts?

![CleanShot 2024-01-03 at 15 01 41](https://github.com/elastic/kibana/assets/3884767/4010737b-d62a-41c8-9cc5-3145622dcf63)

<hr/>

I know we have new gauge visualization styles coming soon from @gvnmagni, but I wanted to point out that the reduced workspace size enforced by the aspect ratio for horizontal gauges exacerbates the feeling that the visualization is not visually centered when you have a short title. I'm assuming this is because the current gauge visualization styles are giving equal width to the gauge title and gauge chart. Is there anything we can do in the mean time to address this? Or should we just wait for those upcoming changes?

![CleanShot 2024-01-03 at 15 39 16](https://github.com/elastic/kibana/assets/3884767/df9bcd48-165c-4f9b-b662-f05289fd44d9)

drewdaemon commented on src/plugins/chart_expressions/common/dimension_transition_veil.tsx:
```diff
@@ -0,0 +1,62 @@
+  return {
+    veil: (
+      <div
+        css={{
+          height: '100%',
+          width: '100%',
+          backgroundColor: 'white',
+          position: 'absolute',
+          zIndex: 1,
+          display: showVeil ? 'block' : 'none',
+        }}
```
This is using emotion. Just the [object syntax](https://emotion.sh/docs/css-prop#object-styles).

drewdaemon commented:
@MichaelMarcialis 

> When shrinking your browser viewport width while using a vertical bar visualization, just prior to hitting the small breakpoint threshold, I noticed the workspace isn't shrinking horizontally and is instead overflowing horizontally out of its container. Can we fix this?

The [CSS aspect-ratio](https://developer.mozilla.org/en-US/docs/Web/CSS/aspect-ratio) prop requires that at least one dimension be automatically sized. You can see [here](https://github.com/elastic/kibana/pull/168651/files#diff-4b4e64e5b7396b5296dac35c3012aaeb2197e7bc613ba5c4b27ef788c73fd41cR63-R71) that the code chooses the largest dimension to constrain (to `100%`). In the case of a horizontal bar, that will be the height not the width.

In order for us to accomplish what you're asking for here I think we'd have to be a bit smarted and detect the smallest dimension of the _workspace_. I can give this a try.

> It appears that a square aspect ratio is being applied to table visualizations. Is this intentional? I think we should have tables always expand to the full available width and height in the workspace. Thoughts?

This isn't supposed to happen. Sometimes the dimensions aren't correctly set during a chart transition. My guess is that you transitioned from metric to table but the workspace kept its 300x300 dimensions.

I mentioned this under my known issues list—still looking for a fix

stratoula commented:
> I know we have new gauge visualization styles coming soon from @gvnmagni, but I wanted to point out that the reduced workspace size enforced by the aspect ratio for horizontal gauges exacerbates the feeling that the visualization is not visually centered when you have a short title.

@MichaelMarcialis @drewdaemon we can do 2 things here:

- Do not apply the aspect ratio now, but do it when the new design lands to be sure that this is going to look fine in the same design
- If we are confident that with the new design the problem is solved then we can do it now. Gauges are in tech preview so we can do this kind of stuff as long as we are sure that the new design is going to look fine with the new ratio. 

I am ok with both options. I leave it to you to decide

stratoula commented:
> Good question. I didn't consider changing the dimensions based on anything but chart type, but I'm sure we could. Not sure where to discuss this.

I will add it as a discuss item on our Monday's sync. I want @timductive to chime in for this!

mbondyra reviewed (approved):
I feel like the horizontal bar issue doesn't belong to this PR so approving! 
Regarding keeping the ratio, the xy charts are still changing the ratio slightly when resizing, but I do actually like it how it works. 

https://github.com/elastic/kibana/assets/4283304/6f7d381b-3491-4677-8996-1168f0f4a99a

drewdaemon commented:
@elasticmachine merge upstream

MichaelMarcialis commented:
> > When shrinking your browser viewport width while using a vertical bar visualization, just prior to hitting the small breakpoint threshold, I noticed the workspace isn't shrinking horizontally and is instead overflowing horizontally out of its container. Can we fix this?
> 
> The [CSS aspect-ratio](https://developer.mozilla.org/en-US/docs/Web/CSS/aspect-ratio) prop requires that at least one dimension be automatically sized. You can see [here](https://github.com/elastic/kibana/pull/168651/files#diff-4b4e64e5b7396b5296dac35c3012aaeb2197e7bc613ba5c4b27ef788c73fd41cR63-R71) that the code chooses the largest dimension to constrain (to `100%`). In the case of a horizontal bar, that will be the height not the width.

@drewdaemon: Thanks for those details! I did some further investigation, and it looks like this issue can be mitigated by placing a `max-width` (and/or `max-height` to account for other visualization types) of 100% on the element getting the `aspect-ratio` style. Possible to add this fix?

![CleanShot 2024-01-04 at 15 30 18](https://github.com/elastic/kibana/assets/3884767/d9e251fe-8ee5-40bf-a46c-5384f9d51f56)

> > It appears that a square aspect ratio is being applied to table visualizations. Is this intentional? I think we should have tables always expand to the full available width and height in the workspace. Thoughts?
> 
> This isn't supposed to happen. Sometimes the dimensions aren't correctly set during a chart transition. My guess is that you transitioned from metric to table but the workspace kept its 300x300 dimensions.

@drewdaemon: I'm not sure that it's the transitioning issue, as I was coming from a stacked bar visualization and it appears to consistently happen as I shrink the viewport width with a table visualization. In looking at the styles, it looks like these two styles may be the culprits. Returning the `aspect-ratio` to auto and giving it a `height` of 100% (or a `flex` of 1) fixes the problem.

![CleanShot 2024-01-04 at 15 18 44](https://github.com/elastic/kibana/assets/3884767/9713819b-ef27-4173-9ef1-e4dbeba85ca0)



> > I know we have new gauge visualization styles coming soon from @gvnmagni, but I wanted to point out that the reduced workspace size enforced by the aspect ratio for horizontal gauges exacerbates the feeling that the visualization is not visually centered when you have a short title.
> 
> @MichaelMarcialis @drewdaemon we can do 2 things here:
> 
> * Do not apply the aspect ratio now, but do it when the new design lands to be sure that this is going to look fine in the same design
> * If we are confident that with the new design the problem is solved then we can do it now. Gauges are in tech preview so we can do this kind of stuff as long as we are sure that the new design is going to look fine with the new ratio.
> 
> I am ok with both options. I leave it to you to decide

@stratoula: I don't see this particular issue as a hard blocker, personally. Yes, it looks visually off-center, but it doesn't render the visualization unusable. 

I'd argue the exacerbating of the issue with lengthy labels and/or titles causing the visualization to be hidden may be more of a reason to pause our applying these aspect ratios. But again, as I mentioned previously, it's technically an issue we already have today as well; it's just not seen as frequently because we're not currently constraining the workspace dimensions. Perhaps it something we can discuss with @gvnmagni and @markov00 during the next visualization sync, as maybe there is a solution on the Elastic charts side of things.

drewdaemon commented:
@MichaelMarcialis 

> @drewdaemon: Thanks for those details! I did some further investigation, and it looks like this issue can be mitigated by placing a max-width (and/or max-height to account for other visualization types) of 100% on the element getting the aspect-ratio style. Possible to add this fix?

Thanks for looking into this! But as far as I know I had fixed this case before your last comment. Could you double-check?

> @drewdaemon: I'm not sure that it's the transitioning issue, as I was coming from a stacked bar visualization and it appears to consistently happen as I shrink the viewport width with a table visualization. In looking at the styles, it looks like these two styles may be the culprits. Returning the aspect-ratio to auto and giving it a height of 100% (or a flex of 1) fixes the problem.

I think I had fixed this as well. Could you double-check that it has been resolved?

drewdaemon commented:
> I'd argue the exacerbating of the issue with lengthy labels and/or titles causing the visualization to be hidden may be more of a reason to pause our applying these aspect ratios. But again, as I mentioned previously, it's technically an issue we already have today as well; it's just not seen as frequently because we're not currently constraining the workspace dimensions. Perhaps it something we can discuss with @gvnmagni and @markov00 during the next visualization sync, as maybe there is a solution on the Elastic charts side of things.

I agree. This is potentially a blocker. We'll discuss then 👍

drewdaemon commented:
@elasticmachine merge upstream

drewdaemon commented:
@stratoula tests don't seem flaky anymore: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4755

MichaelMarcialis commented:
> > @drewdaemon: Thanks for those details! I did some further investigation, and it looks like this issue can be mitigated by placing a max-width (and/or max-height to account for other visualization types) of 100% on the element getting the aspect-ratio style. Possible to add this fix?
> 
> Thanks for looking into this! But as far as I know I had fixed this case before your last comment. Could you double-check?
> 
> > @drewdaemon: I'm not sure that it's the transitioning issue, as I was coming from a stacked bar visualization and it appears to consistently happen as I shrink the viewport width with a table visualization. In looking at the styles, it looks like these two styles may be the culprits. Returning the aspect-ratio to auto and giving it a height of 100% (or a flex of 1) fixes the problem.
> 
> I think you're right—I discovered this as a different bug. I think this is fixed as well. Could you double-check that it has been resolved?

Thanks, @drewdaemon! You're right, both of those have already been corrected by your updates. I hadn't yet pulled in your latest changes when replying. Both look good now.

stratoula commented:
@drewdaemon amazing, let's figure out the final details today on our sync and I will re-test and approve!

drewdaemon commented:
## Current status of this PR

1. During a team meeting we determined to make two changes
    - [x] an XY chart should take the total space available unless it has a time-based X axis
    - [x] metric and gauge charts should _not_ retain their aspect ratio when there isn't enough room in the workspace to accommodate the size request in either or both directions

2. During a subsequent developer sync the following question was raised: should a time-based XY chart retain a 16:9 aspect ratio or simply enforce a minimum height. Waiting for feedback from @gvnmagni on this particular question.

3. One final question for @gvnmagni or @MichaelMarcialis — do we like this small white border on the outside of the workspace or should it be removed?

<img width="641" alt="Screenshot 2024-01-09 at 8 00 44 AM" src="https://github.com/elastic/kibana/assets/315764/83a6eed6-2bdd-4836-9a3c-2fbeb24a67f5">

markov00 commented:
@drewdaemon I've discussed this with @gvnmagni and we are finally aligned on the time-series aspect ratio.
**It should be 16/9, always taking the full available panel width.**
The height instead will follow these rules:
- the height can't be higher than the max available panel height: so basically with very large screens 16/9 is not fully respected because the width is always full width, but the height can extend to the max screen height.
- the height can't be smaller than 300px (arbitrary choosed) to avoid reducing the panel too much. Down to a width of ~550px everything respects the 16/9, after then it becomes more and more "squarish" as it reduces

drewdaemon commented:
@elasticmachine merge upstream

drewdaemon commented:
@markov00 @gvnmagni have I made your wildest dreams come true?

https://github.com/elastic/kibana/assets/315764/b7812097-d5fc-4ee3-977b-21e399d5cf63

gvnmagni commented:
> @markov00 @gvnmagni have I made your wildest dreams come true?
> 
>  Screen.Recording.2024-01-10.at.2.26.10.PM.mov

This is perfect! 😍

markov00 commented:
> @markov00 @gvnmagni have I made your wildest dreams come true?

Perfect thanks!

MichaelMarcialis commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/workspace_panel/workspace_panel_wrapper.scss:
```diff
@@ -26,7 +26,7 @@
     height: 100%;
 
-    > * {
+    >* {
```
Small formatting suggestion.

```suggestion
    & > * {
```

MichaelMarcialis reviewed (approved):
Thanks again for making those earlier fixes, @drewdaemon. I've left a few small comments for your re-review, but nothing worth holding you up over. Assuming you're able to address them, I'll approve now.

drewdaemon commented on src/plugins/chart_expressions/common/chart_size_transition_veil.tsx:
```diff
@@ -0,0 +1,74 @@
+        css={{
+          height: '100%',
+          width: '100%',
+          backgroundColor: 'white',
```
Very good catch 😅

drewdaemon commented:
@elasticmachine merge upstream

drewdaemon commented:
Flaky test runner passed again https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/4859#018cffae-0178-4df7-8091-c75e60d3299e
